src/platform/nrf_pca10001/lib.rs

```diff
@@ -46,70 +46,79 @@ impl Firestorm {
 }
 
 macro_rules! static_init {
-   ($V:ident : $T:ty = $e:expr) => {
-        let $V : &mut $T = {
-            // Waiting out for size_of to be available at compile-time to avoid
-            // hardcoding an abitrary large size...
-            static mut BUF : [u8; 1024] = [0; 1024];
+    ($V:ident : $T:ty = $e:expr, $size:expr) => {
+        // Ideally we could use mem::size_of<$T> here instead of $size, however
+        // that is not currently possible in rust. Instead we write the size as
+        // a constant in the code and use compile-time verification to see that
+        // we got it right
+        let $V : &'static mut $T = {
+            // This is our compile-time assertion. The optimizer should be able
+            // to remove it from the generated code.
+            let assert_buf: [u8; $size] = [0; $size];
+            let assert_val: $T = mem::transmute(assert_buf);
+            mem::forget(assert_val);
+
+            // Statically allocate a read-write buffer for the value, write our
+            // initial value into it (without dropping the initial zeros) and
+            // return a reference to it.
+            static mut BUF: [u8; $size] = [0; $size];
             let mut tmp : &mut $T = mem::transmute(&mut BUF);
-            *tmp = $e;
+            ptr::write(tmp as *mut $T, $e);
             tmp
         };
-   }
+    }
 }
 
-pub unsafe fn init<'a>() -> &'a mut Firestorm {
-    use core::mem;
+pub unsafe fn init() -> &'static mut Firestorm {
+    use core::{mem, ptr};
```

Comment:
hrm... since `ptr` isn't used anywhere explicitly in this function (rather only inside of `static_init!`) maybe it should be imported within a scope there? Just would be less confusing IMO